### DISCUSSION TITLE
[Question] What does fluss need to be configured to support Minio's S3?

### DISCUSSION THREAD

**Discussion by biao-lvwan** [apache/fluss#214] [Q&A] [answered]
### Search before asking

- [X] I searched in the [issues](https://github.com/alibaba/fluss/issues) and found nothing similar.

### Fluss version

0.5.0

### Minimal reproduce step

1，fluss config  remote.data.dir: s3://fluss/storage, appear Region is not set
2，user lakehouse.sh    Dynamic session credentials for Fluss: No AWS Credentials

### What doesn't meet your expectations?

1,The data can be written to remote.data.dir, but an error message will appear and the Region is not set
2,use lakehouse.sh ,Can't read data from Flink's S3 configuration? Where do I need to configure it specifically,Can't I read the S3 configuration configured in Flink

### Anything else?

<img width="725" alt="bd2cc512cd7fd39dbc8d78eb6ff12f8" src="https://github.com/user-attachments/assets/b7313e8f-7ff4-42ac-ad98-4596e758464f" />
<img width="911" alt="9dec068fd31fbbc15b1c0f65265aa2a" src="https://github.com/user-attachments/assets/979ef0d8-4cf7-4550-9437-0067550bc735" />
![image](https://github.com/user-attachments/assets/1d1e93b6-52b4-4cc6-9008-8cffe0221d0b)

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR!

**Comment by luoyuxia** [comment]
@biao-lvwan Thanks for reporting!  You'll need to set `s3.region` in Fluss's `server.yaml` to solve the excpetion as the [doc](https://alibaba.github.io/fluss-docs/docs/maintenance/filesystems/s3/) said . For the region, you can just use `us-east-1` as the region of  Minio's s3...
But I have to remind you that when reading from s3, Fluss client requires to obtain STS token from S3 by the method:
```
 AWSSecurityTokenService stsClient =
                AWSSecurityTokenServiceClientBuilder.standard()
                        .withRegion(region)
                        .withCredentials(
                                new AWSStaticCredentialsProvider(
                                        new BasicAWSCredentials(accessKey, secretKey)))
                        .build();
```
But unfortunately, it's not supported by Minio's s3.......
So you may still see a exception  like `Failed to update tokens ...` in client side, but I believe the exception won't cause the client read fail..

**Comment by biao-lvwan** [comment]
> @biao-lvwan Thanks for reporting! You'll need to set in Fluss's to solve the excpetion as the [doc](https://alibaba.github.io/fluss-docs/docs/maintenance/filesystems/s3/) said . But I have to remind you that when reading from s3, Fluss client requires to obtain STS token from S3 by the method:`s3.region``server.yaml`
> 
> ```
>  AWSSecurityTokenService stsClient =
>                 AWSSecurityTokenServiceClientBuilder.standard()
>                         .withRegion(region)
>                         .withCredentials(
>                                 new AWSStaticCredentialsProvider(
>                                         new BasicAWSCredentials(accessKey, secretKey)))
>                         .build();
> ```
> 
> But unfortunately, it's not supported by Minio's s3....... So you may see a exception like in client side, but I believe the exception won't cause the client read fail..`Failed to update tokens ...`

<img width="638" alt="e6e545d471827e34703a7bed6c0fdba" src="https://github.com/user-attachments/assets/47ecfc2a-87dd-4e1f-8520-8fa38c440e12" />
<img width="638" alt="e6e545d471827e34703a7bed6c0fdba" src="https://github.com/user-attachments/assets/f2f5f6d8-1c45-4089-a3ad-a256322c1259" />

Yes, this error does not affect the remote data transfer to Minio, but when opening lakehouse.sh, Start to download kv snapshot files to local directory for bucket， This task has been submitted to the Flink cluster, and after startup, insert，tm, It will report an error that the configuration has not been obtained，Do I need to configure server. yaml in the Flink cluster?

**Comment by luoyuxia** [comment, marked as answer]
> > @biao-lvwan Thanks for reporting! You'll need to set in Fluss's to solve the excpetion as the [doc](https://alibaba.github.io/fluss-docs/docs/maintenance/filesystems/s3/) said . But I have to remind you that when reading from s3, Fluss client requires to obtain STS token from S3 by the method:`` s3.region``server.yaml ``
> > ```
> >  AWSSecurityTokenService stsClient =
> >                 AWSSecurityTokenServiceClientBuilder.standard()
> >                         .withRegion(region)
> >                         .withCredentials(
> >                                 new AWSStaticCredentialsProvider(
> >                                         new BasicAWSCredentials(accessKey, secretKey)))
> >                         .build();
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > But unfortunately, it's not supported by Minio's s3....... So you may see a exception like in client side, but I believe the exception won't cause the client read fail..`Failed to update tokens ...`
> 
> <img alt="e6e545d471827e34703a7bed6c0fdba" width="638" src="https://private-user-images.githubusercontent.com/66588660/396461430-47ecfc2a-87dd-4e1f-8520-8fa38c440e12.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQ0MzQwOTAsIm5iZiI6MTczNDQzMzc5MCwicGF0aCI6Ii82NjU4ODY2MC8zOTY0NjE0MzAtNDdlY2ZjMmEtODdkZC00ZTFmLTg1MjAtOGZhMzhjNDQwZTEyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMTclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjE3VDExMDk1MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTVlZmE0MzlkMzAyYmNkOGRmODE0ZTEwZmM0MGE0NWFjYWVkM2U2MWZkZDg3ZDMyNjU2N2I1NmZhYTVmMjlkY2UmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.EJERGKMCetxA8XLEeslHZVmFDXjiV7-mFxGzUJDFqEI"> <img alt="e6e545d471827e34703a7bed6c0fdba" width="638" src="https://private-user-images.githubusercontent.com/66588660/396461509-f2f5f6d8-1c45-4089-a3ad-a256322c1259.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQ0MzQwOTAsIm5iZiI6MTczNDQzMzc5MCwicGF0aCI6Ii82NjU4ODY2MC8zOTY0NjE1MDktZjJmNWY2ZDgtMWM0NS00MDg5LWEzYWQtYTI1NjMyMmMxMjU5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMTclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjE3VDExMDk1MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTk0NDc4MjI1YzBiNmYzNjU5ODRhODFjN2QyNjA2YTAwZWNjNjVlMWFiNjY4YTUzOTBkZmI0OTVjM2M1M2QzYWQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.IWamNXFebwdD56GWeGQdsnCuFoLObtTVpqHCnFtbyTY">
> Yes, this error does not affect the remote data transfer to Minio, but when opening lakehouse.sh, Start to download kv snapshot files to local directory for bucket， This task has been submitted to the Flink cluster, and after startup, insert，tm, It will report an error that the configuration has not been obtained，Do I need to configure server. yaml in the Flink cluster?

Thanks for explantion. Now, I got you.. Unfortunately you can't read with Minio's S3 since Fluss client can't get the a credential token from your Minio's S3 for the reason I said before. Fluss will use S3's api to obtain the credential token, but Minio don't support such API..

**Comment by biao-lvwan** [comment]
> > > @biao-lvwan Thanks for reporting! You'll need to set in Fluss's to solve the excpetion as the [doc](https://alibaba.github.io/fluss-docs/docs/maintenance/filesystems/s3/) said . But I have to remind you that when reading from s3, Fluss client requires to obtain STS token from S3 by the method:server.yaml ``` s3.region`
> > > ```
> > >  AWSSecurityTokenService stsClient =
> > >                 AWSSecurityTokenServiceClientBuilder.standard()
> > >                         .withRegion(region)
> > >                         .withCredentials(
> > >                                 new AWSStaticCredentialsProvider(
> > >                                         new BasicAWSCredentials(accessKey, secretKey)))
> > >                         .build();
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > But unfortunately, it's not supported by Minio's s3....... So you may see a exception like in client side, but I believe the exception won't cause the client read fail..`Failed to update tokens ...`
> > 
> > 
> > <img alt="e6e545d471827e34703a7bed6c0fdba" width="638" src="https://private-user-images.githubusercontent.com/66588660/396461430-47ecfc2a-87dd-4e1f-8520-8fa38c440e12.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQ0MzQwOTAsIm5iZiI6MTczNDQzMzc5MCwicGF0aCI6Ii82NjU4ODY2MC8zOTY0NjE0MzAtNDdlY2ZjMmEtODdkZC00ZTFmLTg1MjAtOGZhMzhjNDQwZTEyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMTclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjE3VDExMDk1MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTVlZmE0MzlkMzAyYmNkOGRmODE0ZTEwZmM0MGE0NWFjYWVkM2U2MWZkZDg3ZDMyNjU2N2I1NmZhYTVmMjlkY2UmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.EJERGKMCetxA8XLEeslHZVmFDXjiV7-mFxGzUJDFqEI"> <img alt="e6e545d471827e34703a7bed6c0fdba" width="638" src="https://private-user-images.githubusercontent.com/66588660/396461509-f2f5f6d8-1c45-4089-a3ad-a256322c1259.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQ0MzQwOTAsIm5iZiI6MTczNDQzMzc5MCwicGF0aCI6Ii82NjU4ODY2MC8zOTY0NjE1MDktZjJmNWY2ZDgtMWM0NS00MDg5LWEzYWQtYTI1NjMyMmMxMjU5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMTclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjE3VDExMDk1MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTk0NDc4MjI1YzBiNmYzNjU5ODRhODFjN2QyNjA2YTAwZWNjNjVlMWFiNjY4YTUzOTBkZmI0OTVjM2M1M2QzYWQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.IWamNXFebwdD56GWeGQdsnCuFoLObtTVpqHCnFtbyTY">
> > Yes, this error does not affect the remote data transfer to Minio, but when opening lakehouse.sh, Start to download kv snapshot files to local directory for bucket， This task has been submitted to the Flink cluster, and after startup, insert，tm, It will report an error that the configuration has not been obtained，Do I need to configure server. yaml in the Flink cluster?
> 
> Thanks for explantion. Now, I got you.. Unfortunately you can't read with Minio's S3 since Fluss client can't get the a credential token from your Minio's S3 for the reason I said before. Fluss will use S3's api to obtain the credential token, but Minio don't support such API..

thx

**Comment by luoyuxia** [comment]
But I think maybe Fluss can use the API that Minio also supports to obtain the token like [Trino](https://trino.io/docs/current/object-storage/file-system-s3.html) does. Seems trino use  leverage sts endpoint to do that..

**Comment by biao-lvwan** [comment]
> But I think maybe Fluss can use the API that Minio also supports to obtain the token like [Trino](https://trino.io/docs/current/object-storage/file-system-s3.html) does. Seems trino use leverage sts endpoint to do that..

Thank you, I'll give it a try
